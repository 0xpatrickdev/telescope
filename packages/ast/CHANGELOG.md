# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.3.9](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.3.8...@osmonauts/ast@0.3.9) (2022-05-18)

**Note:** Version bump only for package @osmonauts/ast





## [0.3.8](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.3.7...@osmonauts/ast@0.3.8) (2022-05-18)

**Note:** Version bump only for package @osmonauts/ast





## [0.3.7](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.3.6...@osmonauts/ast@0.3.7) (2022-05-17)

**Note:** Version bump only for package @osmonauts/ast





## [0.3.6](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.3.5...@osmonauts/ast@0.3.6) (2022-05-17)

**Note:** Version bump only for package @osmonauts/ast





## [0.3.5](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.3.4...@osmonauts/ast@0.3.5) (2022-05-14)

**Note:** Version bump only for package @osmonauts/ast





## [0.3.4](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.3.3...@osmonauts/ast@0.3.4) (2022-05-14)

**Note:** Version bump only for package @osmonauts/ast





## [0.3.3](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.3.2...@osmonauts/ast@0.3.3) (2022-05-14)

**Note:** Version bump only for package @osmonauts/ast





## [0.3.2](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.3.1...@osmonauts/ast@0.3.2) (2022-05-14)

**Note:** Version bump only for package @osmonauts/ast





## [0.3.1](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.3.0...@osmonauts/ast@0.3.1) (2022-05-14)

**Note:** Version bump only for package @osmonauts/ast





# [0.3.0](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast@0.2.1...@osmonauts/ast@0.3.0) (2022-05-14)

**Note:** Version bump only for package @osmonauts/ast





## 0.2.1 (2022-04-28)

**Note:** Version bump only for package @osmonauts/ast





# [0.2.0](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.1.7...@osmonauts/ast-gen@0.2.0) (2022-04-22)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.1.7](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.1.6...@osmonauts/ast-gen@0.1.7) (2022-04-22)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.1.6](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.1.5...@osmonauts/ast-gen@0.1.6) (2022-04-22)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.1.5](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.1.4...@osmonauts/ast-gen@0.1.5) (2022-04-21)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.1.4](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.1.3...@osmonauts/ast-gen@0.1.4) (2022-04-21)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.1.3](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.1.2...@osmonauts/ast-gen@0.1.3) (2022-04-21)

**Note:** Version bump only for package @osmonauts/ast-gen





## 0.1.2 (2022-04-21)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.1.1](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.1.0...@osmonauts/ast-gen@0.1.1) (2022-04-18)

**Note:** Version bump only for package @osmonauts/ast-gen





# 0.1.0 (2022-04-18)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.0.16](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.0.15...@osmonauts/ast-gen@0.0.16) (2022-04-15)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.0.15](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.0.14...@osmonauts/ast-gen@0.0.15) (2022-04-13)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.0.14](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.0.13...@osmonauts/ast-gen@0.0.14) (2022-04-12)

**Note:** Version bump only for package @osmonauts/ast-gen





## [0.0.13](https://github.com/osmosis-labs/telescope/compare/@osmonauts/ast-gen@0.0.12...@osmonauts/ast-gen@0.0.13) (2022-04-12)

**Note:** Version bump only for package @osmonauts/ast-gen
