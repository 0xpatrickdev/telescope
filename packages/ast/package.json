{
  "name": "@osmonauts/ast",
  "version": "0.28.0",
  "description": "Cosmos TypeScript AST generation",
  "author": "Dan Lynch <pyramation@gmail.com>",
  "homepage": "https://github.com/osmosis-labs/telescope/tree/master/packages/ast#readme",
  "license": "SEE LICENSE IN LICENSE",
  "main": "main/index.js",
  "typings": "types/src/index.d.ts",
  "directories": {
    "lib": "src",
    "test": "__tests__"
  },
  "files": [
    "types",
    "main"
  ],
  "scripts": {
    "build": "cross-env BABEL_ENV=production babel src --out-dir main --delete-dir-on-start --extensions \".tsx,.ts,.js\"",
    "build:ts": "tsc --project ./tsconfig.json",
    "buidl": "npm run build && npm run build:ts",
    "prepare": "npm run build",
    "lint": "eslint .",
    "format": "eslint . --fix",
    "test": "jest",
    "test:ast": "cross-env NODE_ENV=development babel-node scripts/test-ast.js",
    "test:watch": "jest --watch",
    "test:debug": "node --inspect node_modules/.bin/jest --runInBand"
  },
  "jest": {
    "testPathIgnorePatterns": [
      "main/",
      "module/",
      "types/"
    ]
  },
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/osmosis-labs/telescope"
  },
  "keywords": [],
  "bugs": {
    "url": "https://github.com/osmosis-labs/telescope/issues"
  },
  "devDependencies": {
    "@babel/cli": "7.18.10",
    "@babel/core": "7.18.13",
    "@babel/eslint-parser": "^7.18.9",
    "@babel/generator": "7.18.13",
    "@babel/node": "^7.18.10",
    "@babel/parser": "^7.18.13",
    "@babel/plugin-proposal-class-properties": "7.18.6",
    "@babel/plugin-proposal-export-default-from": "7.18.10",
    "@babel/plugin-proposal-object-rest-spread": "7.18.9",
    "@babel/plugin-transform-runtime": "7.18.10",
    "@babel/preset-env": "7.18.10",
    "@babel/preset-typescript": "^7.17.12",
    "@babel/traverse": "7.18.13",
    "@types/jest": "^28.1.7",
    "ast-stringify": "0.1.0",
    "babel-core": "7.0.0-bridge.0",
    "babel-jest": "28.1.3",
    "babel-watch": "^7.0.0",
    "cross-env": "^7.0.2",
    "deepmerge": "4.2.2",
    "eslint": "8.22.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-plugin-prettier": "^4.0.0",
    "glob": "8.0.3",
    "jest": "^28.1.3",
    "jest-in-case": "^1.0.2",
    "mkdirp": "1.0.4",
    "prettier": "^2.7.0",
    "regenerator-runtime": "^0.13.7",
    "ts-jest": "^28.0.8",
    "typescript": "^4.7.3"
  },
  "dependencies": {
    "@babel/runtime": "^7.18.9",
    "@babel/types": "7.18.13",
    "@osmonauts/proto-parser": "^0.19.1",
    "@osmonauts/types": "^0.15.1",
    "@osmonauts/utils": "^0.4.0",
    "case": "1.6.3",
    "dotty": "0.1.2"
  }
}
