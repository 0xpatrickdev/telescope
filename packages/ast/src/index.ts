export * from './encoding';
export * from './clients';
export * from './registry';
export * from './services';
export * from './utils';
export * from './bundle';
export * from './types';