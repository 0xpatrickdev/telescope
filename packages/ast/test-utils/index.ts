import generate from '@babel/generator';
import { ProtoStore, getNestedProto, parseProto } from '@osmonauts/proto-parser'
import { ProtoType } from '../src/encoding/proto/types';
import { ParseContext } from '../src/encoding/context';

export const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}
export const printCode = (ast) => {
    console.log(
        generate(ast).code
    );
}

export const prepareContext = (store: ProtoStore, protoFile: string) => {
    const ref = store.findProto(protoFile);
    const traversed = ref.traversed
    if (!traversed) {
        throw new Error('not traversed!');
    }
    const proto: Record<string, ProtoType> = getNestedProto(traversed);
    const protos: ProtoType[] = Object.values(proto).filter(
        proto => proto.name.startsWith('Msg')
            &&
            !proto.name.endsWith('Response')
            &&
            proto.name !== 'Msg'
    );

    const context = new ParseContext(ref, store);

    return {
        context,
        root: traversed,
        protos
    };
};
