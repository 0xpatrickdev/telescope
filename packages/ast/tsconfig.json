{
  "compilerOptions": {
    "baseUrl": ".",
    "rootDir": "src",
    "skipLibCheck": true,
    "declaration": true,
    "declarationDir": "./types",
    "emitDeclarationOnly": false,
    "isolatedModules": true,
    "allowJs": true,

    "esModuleInterop": true,
    "target": "ESNext",
    "lib": ["es2022", "DOM"],
    "downlevelIteration": true,
    "moduleResolution": "node",
    "resolveJsonModule": true
  },
  "include": ["src/**/*"],
  "exclude": [
    "node_modules",
    "test-utils",
    "src/**/*.spec.ts",
    "src/**/*.test.ts",
    "src/__fixtures__/**/*",
    "types"
  ]
}
