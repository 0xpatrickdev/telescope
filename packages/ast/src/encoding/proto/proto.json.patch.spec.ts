import {
    createProtoEnum,
    createProtoEnumFromJSON,
    createProtoEnumToJSON,
} from "./enums";
import {
    getNestedProto,
    convertProtoPathToNestedJSONPath as convert,
    convertPackageNameToNestedJSONPath as convertPackage
} from "@cosmology/utils";
import { ProtoParseContext } from "@cosmology/ast";
import {
    expectCode,
    getTestProtoStore
} from "../../../test-utils";

const p = convert('cosmwasm/wasm/v1/types.proto');
const cp = convertPackage('cosmwasm.wasm.v1');
const store = getTestProtoStore({
    prototypes: {
        patch: {
            'cosmwasm/wasm/v1/types.proto': [
                { op: 'replace', path: '~/AccessType/valuesOptions/ACCESS_TYPE_NOBODY/(gogoproto.enumvalue_customname)', value: 'NobodyAccess' },
                { op: 'replace', path: '@/AccessType/valuesOptions/ACCESS_TYPE_UNSPECIFIED/(gogoproto.enumvalue_customname)', value: 'UnspecifiedAccess' },
                { op: 'replace', path: p + '/AccessType/valuesOptions/ACCESS_TYPE_EVERYBODY/(gogoproto.enumvalue_customname)', value: 'EverybodyAccess' },
                { op: 'replace', path: cp + '/AccessType/valuesOptions/ACCESS_TYPE_ONLY_ADDRESS/(gogoproto.enumvalue_customname)', value: 'OnlyAddressAccess' },
                {
                    op: "add",
                    path: "/root/nested/cosmwasm/nested/wasm/nested/v1/nested/AccessType/values/ACCESS_TYPE_SUPER_FUN",
                    value: 4
                },
                {
                    op: "add",
                    path: "/root/nested/cosmwasm/nested/wasm/nested/v1/nested/AccessType/valuesOptions/ACCESS_TYPE_SUPER_FUN",
                    value: {
                        "(gogoproto.enumvalue_customname)": "SuperFunAccessType"
                    }
                }
            ]
        }
    },
    enums: {
        useCustomNames: true
    }
});
store.traverseAll();

it("createProtoEnum", async () => {
    const ref = store.findProto('cosmwasm/wasm/v1/types.proto');
    const context = new ProtoParseContext(ref, store, store.options);
    expectCode(
        createProtoEnum(
            context,
            "NullValue",
            getNestedProto(ref.traversed).AccessType
        )
    );
});

it("createProtoEnumFromJSON", async () => {
    const ref = store.findProto('cosmwasm/wasm/v1/types.proto');
    const context = new ProtoParseContext(ref, store, store.options);
    expectCode(
        createProtoEnumFromJSON(
            context,
            "AccessType",
            getNestedProto(ref.traversed).AccessType
        )
    );
});

it("createProtoEnumToJSON", async () => {
    const ref = store.findProto('cosmwasm/wasm/v1/types.proto');
    const context = new ProtoParseContext(ref, store, store.options);
    expectCode(
        createProtoEnumToJSON(
            context,
            "AccessType",
            getNestedProto(ref.traversed).AccessType
        )
    );
});
