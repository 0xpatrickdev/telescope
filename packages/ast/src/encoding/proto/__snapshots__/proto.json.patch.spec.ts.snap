// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createProtoEnum 1`] = `
"/** AccessType permission types */
export enum NullValue {
  /** ACCESS_TYPE_UNSPECIFIED - AccessTypeUnspecified placeholder for empty value */
  ACCESS_TYPE_UNSPECIFIED = 0,
  /** ACCESS_TYPE_NOBODY - AccessTypeNobody forbidden */
  ACCESS_TYPE_NOBODY = 1,
  /** ACCESS_TYPE_ONLY_ADDRESS - AccessTypeOnlyAddress restricted to an address */
  ACCESS_TYPE_ONLY_ADDRESS = 2,
  /** ACCESS_TYPE_EVERYBODY - AccessTypeEverybody unrestricted */
  ACCESS_TYPE_EVERYBODY = 3,
  ACCESS_TYPE_SUPER_FUN = 4,
  UNRECOGNIZED = -1,
}"
`;

exports[`createProtoEnumFromJSON 1`] = `
"export function accessTypeFromJSON(object: any): AccessType {
  switch (object) {
    case 0:
    case "UnspecifiedAccess":
      return AccessType.ACCESS_TYPE_UNSPECIFIED;
    case 1:
    case "NobodyAccess":
      return AccessType.ACCESS_TYPE_NOBODY;
    case 2:
    case "OnlyAddressAccess":
      return AccessType.ACCESS_TYPE_ONLY_ADDRESS;
    case 3:
    case "EverybodyAccess":
      return AccessType.ACCESS_TYPE_EVERYBODY;
    case 4:
    case "SuperFunAccessType":
      return AccessType.ACCESS_TYPE_SUPER_FUN;
    case -1:
    case "UNRECOGNIZED":
    default:
      return AccessType.UNRECOGNIZED;
  }
}"
`;

exports[`createProtoEnumToJSON 1`] = `
"export function accessTypeToJSON(object: AccessType): string {
  switch (object) {
    case AccessType.ACCESS_TYPE_UNSPECIFIED:
      return "UnspecifiedAccess";
    case AccessType.ACCESS_TYPE_NOBODY:
      return "NobodyAccess";
    case AccessType.ACCESS_TYPE_ONLY_ADDRESS:
      return "OnlyAddressAccess";
    case AccessType.ACCESS_TYPE_EVERYBODY:
      return "EverybodyAccess";
    case AccessType.ACCESS_TYPE_SUPER_FUN:
      return "SuperFunAccessType";
    case AccessType.UNRECOGNIZED:
    default:
      return "UNRECOGNIZED";
  }
}"
`;
